### PR TITLE
SI-6415 avoid early evaluation in some Stream's methods

### PR DESCRIPTION
**Pull request by jedesah**
The lengthCompare method in LinearSeqOptimized was looking one step further
than it needed to in order to give the correct result, which was creating some
unwanted side effects related to Streams.

Add test case to expose the problem and a few test cases to ensure correctness of LinearSeq.lengthCompare() method.

https://issues.scala-lang.org/browse/SI-6415

Review by @axel22

### CODE REVIEW

**Comment by axel22** [comment]
LGTM

**Comment by jedesah** [comment]
As described in the commit message, performed similar modification in the SeqLike implementation.

I wish I knew why I can't remove the method in LinearSeqOptimized. It sorta sucks that it's not a compile time error, is this a bug you think or is it expected behavior?

**Comment by retronym** [comment]
> note: I tried to get rid of the method in LinearSeqOptimized since this one
> will do, but that turned out impossible because of some behaviour and failed
> to understand relative to method overloading.

Can you please elaborate on this?

**Comment by jedesah** [comment]
Ya np, sorry about that, I was clearly a bit tired at that point last night.

The LinearSeqOptimized trait inherits indirectly from the trait SeqLike trait. I tried removing the implementation of lengthCompare present in the LinearSeqOptimized trait because I don't think it adds much and creates code duplication (as far as I can tell they are both O(length min len)), but that turned out impossible du to a runtime error of type NoSuchMethod. For some reason the runtime was not finding the implementation of the function when it was being called.

I thought it might be related to the change that made these traits inherit from Any rather than AnyRef, but that just changed the error from NoSuchMethod to AbstractMethodError. I tried implementing the later with just super.lengthCompare(len), but that didn't work either.

I am using OpenJDK 7 on Fedora to develop, could this be related? I'd be very interested in figuring out what is going on here, I find the behavior odd, but then again, I haven't totally grasped the addition of Value classes and all that this implies yet. In any case, I'm a bit surprised that this is a runtime error rather than a compile time error though, which is what leads me to think this might actually be a bug rather than me not understanding something.

In the mean time, all I have done in this new commit, is change the implementation of the lengthCompare(len) method in SeqLike to reflect the semantic change that I made in LinearSeqOptimized to fix the issue of lazy evaluation at hand.

**Comment by retronym** [comment]
Sounds like (I hope!) an artifact of a Ant's inability to rebuild dependencies. Did you try a clean build? (Or, a little quicker: `find  src/library/collection | xargs touch; ant`

**Comment by retronym** [comment]
But regardless, it's probably best to leave the `while` based implementation in `LinearSeqOptimized`, it could be a touch more efficient.

**Comment by jedesah** [comment]
Unfortunately, it's unlikely to be an artifact of Ant's inability to rebuild dependencies, as I rigorously tested the behavior I described above with full rebuilds to try and understand what was going on.

This might be obvious, but pardon my ignorance, why do think the while based implementation would be more efficient?

**Comment by retronym** [comment]
`for (x <- xs) ...` desugars to `xs.foreach(x => ...)`. Ideally the combination of scalac and HotSpot could make that as fast as a raw while loop, but we're not quite there yet.

Some background reading: http://www.azulsystems.com/blog/cliff/2011-04-04-fixing-the-inlining-problem

Can you please commit the failing code to a branch so I can take a look?

**Comment by jedesah** [comment]
Thx, I'll read it right now.

I could make a branch, but I think that it's not worth the overhead, if you grab this branch and delete the implementation of lengthCompare in LinearSeqOptimized.scala, you should see the behavior I described if it is something that is not particular to my setup. Actually, It would be "unexpected" if it were to work on 2.10.x vanilla based on my observations.

I can create a branch if you really prefer that way though.

**Comment by retronym** [comment]
As someone who chases these things a bit, I can tell you it's a big advantage if you could provide unambiguous steps to reproduce. When did you see that exception? A branch and a transcript of your build/test are the best way to communicate that.

I forgot to say before, that we can't remove a method in a public trait in a minor release. See @jsuereth's slides on [Binary Resiliance](https://docs.google.com/presentation/pub?id=16E5veHv4Kh3w4t-MjkiHSsjXrqjwvH0Lf8rSN-I23mQ&start=false&loop=false&delayms=3000#slide=id.p) to learn about what that means.

But, if removing a method leads to that NSMError or AMError like you've seen, I'm very interested to find out what's happening.

BTW, thanks for this contribution, and your other recent ones!

**Comment by jedesah** [comment]
Good news! It was Ant's (or my) fault after all. I tried again with ant clean.all just to make sure and it ended up working. I might have been using only ant clean before I'm not sure.

Okay, so from here I guess I should implement the two methods uniformly and in a way to maximize lazyness and then submit another pull request to master which removes the implementation in LinearSqOptimized. How does that sound?

**Comment by retronym** [comment]
Excellent.

I would leave the {{LinearSeqOptimized}} implementation, even in master, performance grounds.

(edited to remove bogus source compatibility concerns)

**Comment by paulp** [comment]
"Good news! It was Ant's (or my) fault after all."

You can trust us on that sort of diagnosis; we've been doing this a while. (I wasn't around for the early conversation, but I'd have bet a limb on the eventual resolution.)

**Comment by jedesah** [comment]
"I would leave the {{LinearSeqOptimized}} implementation, even in master, performance grounds."

Okay, no problem.

Should I remove the //TODO then?

To be clear, aside from the //TODO, is this change ready for merge or do I need to change something?

**Comment by adriaanm** [comment]
ping @retronym, @axel22  for a quick follow-up review,

@jedesah, it's best to @-mention people to get your signal to jump out of the github noise

File: src/library/scala/collection/LinearSeqOptimized.scala [hunk 1]
```diff
@@ -247,14 +247,21 @@ trait LinearSeqOptimized[+A, +Repr <: LinearSeqOptimized[A, Repr]] extends Linea
...
 
   override /*SeqLike*/
-  def lengthCompare(len: Int): Int =  {
-    var i = 0
-    var these = self
-    while (!these.isEmpty && i <= len) {
-      i += 1
-      these = these.tail
+  def lengthCompare(len: Int): Int = {
+    // TODO: Remove this method when incrementing a major revision
```
Comment (retronym) [inline src/library/scala/collection/LinearSeqOptimized.scala]: Remove this `TODO`, this implementation looks more efficient that the superclass impl.

File: src/library/scala/collection/LinearSeqOptimized.scala [hunk 2]
```diff
@@ -247,14 +247,21 @@ trait LinearSeqOptimized[+A, +Repr <: LinearSeqOptimized[A, Repr]] extends Linea
...
+    // This method is the same as in SeqLike, no need to redefine it
+    if (len < 0) 1
+    else {
+      var i = 0
+      val it = iterator
+      while (it.hasNext) {
+	if (i == len)
+	  return if (it.hasNext) 1 else 0
+	it.next()
+	i += 1
```
Comment (retronym) [inline src/library/scala/collection/LinearSeqOptimized.scala]: indentation oops.

File: src/library/scala/collection/SeqLike.scala [hunk 3]
```diff
@@ -84,13 +84,18 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
...
-      i += 1
+    if (len < 0) 1
+    else {
+      var i = 0
+      val it = iterator
+      while (it.hasNext) {
+	if (i == len)
+	  return if (it.hasNext) 1 else 0
+	it.next()
+	i += 1
```
Comment (retronym) [inline src/library/scala/collection/SeqLike.scala]: indentation is off here, too. (tabs vs spaces, perhaps?)

**Comment by jedesah** [comment]
Sorry for the delay. Got caught up finishing work stuff.

@retronym Not sure what you mean by "this implementation looks more efficient than the superclass impl", they look exactly the same to me. Thx for pointing out the indentation problems, not sure what happened there.

File: src/library/scala/collection/LinearSeqOptimized.scala [hunk 4]
```diff
@@ -247,14 +247,20 @@ trait LinearSeqOptimized[+A, +Repr <: LinearSeqOptimized[A, Repr]] extends Linea
...
+    else {
+      var i = 0
+      val it = iterator
+      while (it.hasNext) {
+        if (i == len) return if (it.hasNext) 1 else 0
+        it.next()
+        i += 1
+      }
+      i - len
     }
```
Comment (paulp) [inline src/library/scala/collection/LinearSeqOptimized.scala:259]: Sorry I'm late to this party, but this seems pretty suboptimal to me. If I understand the motivation being Stream evaluating one too many elements, that can be addressed without inflicting an iterator on it.

```
  override /*SeqLike*/
  def lengthCompare(len: Int): Int = {
    @tailrec def loop(i: Int, xs: Repr): Int = {
      if (i == len)
        if (xs.isEmpty) 0 else 1
      else if (xs.isEmpty)
        -1
      else
        loop(i + 1, xs.tail)
    }
    if (len < 0) 1
    else loop(0, this)
  }
```

**Comment by paulp** [comment]
The failure is spurious; t2544 is testing for the exact return value of lengthCompare despite there being equivalence classes with 2^31 values apiece.

I also wonder what exactly is supposed to happen when you define a sequence with contradictory information, as in t2544.

```
  object Foo extends Seq[Int] {
    def apply(i: Int) = i
    def length = 4
    def iterator = Iterator(0,1,2,3,4)
  }
```

And I also wonder why these redundant means are all abstract in Seq.

**Comment by paulp** [comment]
Superseded by #1828.

**Pull request closed without merging**